### PR TITLE
engine_rocks: Add ttl properties

### PR DESCRIPTION
**Pull request by Connor1996**
### What problem does this PR solve?

part of https://github.com/tikv/tikv/pull/9595

### What is changed and how it works?

Add TTL properties collector

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 
- None

### CODE REVIEW

File: components/engine_traits/src/util.rs [hunk 1]
```diff
@@ -21,3 +23,30 @@ pub fn check_key_in_range(
...
+    if len < number::U64_SIZE {
+        return Err(Error::Codec(codec::Error::ValueLength));
+    }
+    let mut ts = &value_with_ttl[len - number::U64_SIZE..];
+    Ok(number::decode_u64(&mut ts)?)
+}
+
+pub fn strip_expire_ts(value_with_ttl: &[u8]) -> &[u8] {
+    let len = value_with_ttl.len();
+    &value_with_ttl[..len - number::U64_SIZE]
```
Comment (Little-Wallace) [inline components/engine_traits/src/util.rs:42]: Do we need to check the length and return a `Result`?

**Review by Little-Wallace — COMMENTED** [review]
rest LGTM.

File: components/engine_traits/src/util.rs (hunk 1 above)
Comment (Connor1996) [reply components/engine_traits/src/util.rs:42]: It was used in `value() of iter` which returns &[u8] and it is ensured to be called after `get_expire_ts()`.

Comment (Little-Wallace) [reply components/engine_traits/src/util.rs:42]: OK. But it is a dangerous  design...

File: components/engine_rocks/src/ttl_properties.rs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+    fn finish(&mut self) -> HashMap<Vec<u8>, Vec<u8>> {
+        if self.prop.max_expire_ts == 0 && self.prop.min_expire_ts == 0 {
+            return HashMap::default();
+        }
+        RocksTTLProperties::encode(&self.prop).0
+    }
+}
+
+pub struct TTLPropertiesCollectorFactory {}
```
Comment (BusyJay) [inline components/engine_rocks/src/ttl_properties.rs:99]: In rust, it should be named as `TtlPropertiesCollectorFactory`.

File: components/engine_rocks/src/ttl_properties.rs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        if entry_type != DBEntryType::Put {
+            return;
+        }
+        // only consider data keys
+        if !key.starts_with(keys::DATA_PREFIX_KEY) {
+            return;
+        }
+
+        let expire_ts = get_expire_ts(&value).unwrap_or_else(|_| {
+            error!("unexpected ttl key:{:?}, value:{:?}", key, value);
```
Comment (BusyJay) [inline components/engine_rocks/src/ttl_properties.rs:76]: This probably violates the log redaction requirements.

Comment (Connor1996) [reply components/engine_rocks/src/ttl_properties.rs]: remove the log, and the error will be recorded by metrics in later PR

**Review by BusyJay — APPROVED** [review]

File: components/engine_rocks/src/ttl_properties.rs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+    pub fn decode<T: DecodeProperties>(props: &T) -> Result<TtlProperties> {
+        let mut res = TtlProperties::default();
+        res.max_expire_ts = props.decode_u64(PROP_MAX_EXPIRE_TS)?;
+        res.min_expire_ts = props.decode_u64(PROP_MIN_EXPIRE_TS)?;
+        Ok(res)
+    }
+}
+
+impl TtlPropertiesExt for RocksEngine {
```
Comment (hicqu) [inline components/engine_rocks/src/ttl_properties.rs:34]: Seems this logic is same for `MvccProperties` and `TtlProperties`. Just a question, is it possible to merge them like
```rust
trait PropertiesExt {
    type Properties = X;
    fn get_range_properties_cf(...) -> Reuslt<Vec<(String, Self::Properties)>>;
}
```

File: components/engine_rocks/src/ttl_properties.rs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        }
+        // only consider data keys
+        if !key.starts_with(keys::DATA_PREFIX_KEY) {
+            return;
+        }
+
+        let expire_ts = match get_expire_ts(&value) {
+            Ok(ts) => ts,
+            Err(e) => {
+                error!("failed to get expire ts";
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply components/engine_rocks/src/ttl_properties.rs:78]: Not possible now, but we'll introduce keyspace to support that

File: components/engine_rocks/src/ttl_properties.rs (hunk 4 above)
Comment (Connor1996) [reply components/engine_rocks/src/ttl_properties.rs:34]: Seems yes. I think we can leave it for engine traits' future work.

**Pull request merged**